### PR TITLE
[tree] Handle inconsistent state

### PR DESCRIPTION
**Pull request by svenefftinge**
Signed-off-by: Sven Efftinge <sven.efftinge@typefox.io>

### CODE REVIEW

File: packages/core/src/browser/tree/tree.ts [hunk 1]
```diff
@@ -287,8 +287,22 @@ export class TreeImpl implements Tree {
...
+        } else {
+            return this.getRootNode(node.parent);
+        }
+    }
+
     protected addNode(node: TreeNode | undefined): void {
         if (node) {
+            const root = this.getRootNode(node);
+            if (this.nodes[root.id] && this.nodes[root.id] !== root) {
+                console.info('Child node does not belong to this tree. Resetting root.');
```
Comment (benoitf) [inline packages/core/src/browser/tree/tree.ts]: debug trace ?

**Comment by AlexTugarev** [comment]
I've tested, but couldn't see this being logged. OTOH it doesn't break the normal case.

**Comment by svenefftinge** [comment]
Yes, it is a rare case. But the tree-impl is flawed now, as I can update a child that doesn't belong to the hierarchy. Only in that case, the state should be updated.

File: packages/core/src/browser/tree/tree.ts (hunk 1 above)
Comment (svenefftinge) [reply packages/core/src/browser/tree/tree.ts]: Yes, I can change the level.

**Pull request merged**